integrations/outlook/src/index.ts

```diff
@@ -0,0 +1,252 @@
+    // request sent by Microsoft Graph. Return the token
+    // as plain text with a 200 response
+    // https://docs.microsoft.com/graph/webhooks#notification-endpoint-validation
+    if (req.query) {
+      const queryString = req.query as string
+      const queryParameters = queryString.split('&')
+      const queryParams: { [key: string]: string } = {}
+
+      for (const param of queryParameters) {
+        const [key, value] = param.split('=')
+        if (key && value) {
+          queryParams[key] = value
+        }
+      }
+
+      const validationToken = queryParams.validationToken
+
+      if (validationToken) {
+        var str = decodeURIComponent(validationToken).split('+').join(' ')
+        return {
+          status: 200,
+          headers: {
+            'Content-Type': 'text/plain',
+          },
+          body: str,
+        }
+      }
+    }
+
+    if (!req.body) {
+      logger.warn('Handler received an empty body')
+      return
+    }
+
+    const body = JSON.parse(req.body)
+    if (!body.value) {
+      logger.warn('Missing property value in the body')
+      return
+    }
+
+    body.value.forEach(async (message: any) => {
+      const graphClient = getClient(ctx.configuration)
+      if (message.lifecycleEvent) {
+        logger.info(`Handler received lifecycleEvent `)
+        await graphClient.handleLifecycleEvents(message)
+      } else if (message.resourceData) {
+        const odataId = message.resourceData['@odata.id']
+        const notificationContent = await graphClient.getNotificationContent(odataId)
+
+        let emailbody = _.get(notificationContent, 'body.content')
+        const sender = _.get(notificationContent, 'sender.emailAddress.address')
+
+        if (notificationContent.body?.contentType === 'html') {
+          emailbody = extractMessageFromHtml(emailbody)
+        }
+
+        if (!emailbody) {
+          logger.warn('invalid email body: ', JSON.stringify(message, null, 3))
+          return
+        }
+        if (!sender) {
+          logger.warn('missing sender property: ', JSON.stringify(message, null, 3))
+          return
+        }
+
+        const { conversation } = await client.getOrCreateConversation({
+          channel: 'channel',
+          tags: {
+            'outlook:id': notificationContent.conversationId,
+            'outlook:subject': notificationContent.subject,
+            'outlook:from': JSON.stringify(notificationContent.from),
+            'outlook:toRecipients': JSON.stringify(notificationContent.toRecipients),
+            'outlook:ccRecipients': JSON.stringify(notificationContent.ccRecipients),
+            'outlook:refMessageId': notificationContent.id,
+          },
+        })
+
+        const { user } = await client.getOrCreateUser({
+          tags: {
+            'outlook:id': sender,
+          },
+        })
+
+        await client.createMessage({
+          tags: { 'outlook:id': notificationContent.id },
```

Comment:
how can I access this tag? or how can I get the last message id for a given conversation?